server/src/main/java/org/elasticsearch/index/query/FirstQueryBuilder.java

```diff
@@ -0,0 +1,174 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.query.first.FirstQuery;
+import org.elasticsearch.search.SearchModule;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+/**
+ * First query where a document gets score from the first matching query
+ */
+public class FirstQueryBuilder extends AbstractQueryBuilder<FirstQueryBuilder> {
+    public static final String NAME = "first";
+    private static int maxClauseCount = SearchModule.INDICES_MAX_CLAUSE_COUNT_SETTING.getDefault(Settings.EMPTY);
+
+    private static final ParseField QUERIES_FIELD = new ParseField("queries");
+    private final List<QueryBuilder> queryBuilders;
+
+    private static final ConstructingObjectParser<FirstQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(NAME, false,
+        args -> {
+            FirstQueryBuilder cQueryBuilder = new FirstQueryBuilder((List<QueryBuilder>) args[0]);
+            if (args[1] != null) cQueryBuilder.boost((Float) args[1]);
+            if (args[2] != null) cQueryBuilder.queryName((String) args[2]);
+            return cQueryBuilder;
+        });
+
+    static {
+        PARSER.declareObjectArray(constructorArg(), (p,c) -> parseInnerQueryBuilder(p), QUERIES_FIELD);
+        PARSER.declareFloat(optionalConstructorArg(), AbstractQueryBuilder.BOOST_FIELD);
+        PARSER.declareString(optionalConstructorArg(), AbstractQueryBuilder.NAME_FIELD);
+    }
+
+    public static FirstQueryBuilder fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    /**
+     * Create a new {@link FirstQueryBuilder}
+     *
+     * @param queryBuilders a list of queries
+     */
+    public FirstQueryBuilder(List<QueryBuilder> queryBuilders) {
+        if (queryBuilders == null || queryBuilders.size() == 0) {
+            throw new IllegalArgumentException("[queries] cannot be null or empty!");
+        }
+        if (queryBuilders.size() > maxClauseCount) {
+            throw new IllegalArgumentException("Too many query clauses! Should be less or equal [" +  maxClauseCount + "]!");
```

Comment:
nit: maybe add a test for this exception

Reply:
@cbuescher Thanks for the feedback, I have added a test in `FirstQueryBuilderTests :: testFirstQueryBuilderValidation`